File: backend/app/api/transfers.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright 2026 Lorenzo Benfenati
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.deps import get_current_user, get_house_member
from app.models.house import House
from app.models.user import User
from app.schemas.transfer import (
    AddContainersRequest,
    LocationSummaryOut,
    TransferCreate,
    TransferDetail,
    TransferPlanRequest,
    TransferPlanResponse,
    TransferSummary,
    TransferUpdate,
)
from app.services import transfer_service

router = APIRouter()


@router.get("/houses/{house_id}/transfers", response_model=list[TransferSummary])
async def list_transfers(
    status: str | None = Query(None),
    destination_location_id: str | None = Query(None),
    house: House = Depends(get_house_member),
    db: AsyncSession = Depends(get_db),
):
    return await transfer_service.list_transfers(house.id, db, status, destination_location_id)


@router.post("/houses/{house_id}/transfers", response_model=TransferDetail)
async def create_transfer(
    data: TransferCreate,
    house: House = Depends(get_house_member),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await transfer_service.create_transfer(house.id, current_user.id, data, db)


@router.post("/houses/{house_id}/transfers/plan", response_model=TransferPlanResponse)
async def plan_transfers(
    request: TransferPlanRequest,
    house: House = Depends(get_house_member),
    db: AsyncSession = Depends(get_db),
):
    return await transfer_service.plan_transfers(house.id, request, db)


@router.get("/houses/{house_id}/transfers/{transfer_id}", response_model=TransferDetail)
async def get_transfer(
    transfer_id: str,
    house: House = Depends(get_house_member),
    db: AsyncSession = Depends(get_db),
):
    return await transfer_service.get_transfer_detail(transfer_id, house.id, db)


@router.put("/houses/{house_id}/transfers/{transfer_id}", response_model=TransferDetail)
async def update_transfer(
    transfer_id: str,
    data: TransferUpdate,
    house: House = Depends(get_house_member),
    db: AsyncSession = Depends(get_db),
):
    return await transfer_service.update_transfer(transfer_id, house.id, data, db)


@router.post("/houses/{house_id}/transfers/{transfer_id}/start", response_model=TransferDetail)
async def start_transfer(
    transfer_id: str,
    house: House = Depends(get_house_member),
    db: AsyncSession = Depends(get_db),
):
    return await transfer_service.start_transfer(transfer_id, house.id, db)


@router.post("/houses/{house_id}/transfers/{transfer_id}/complete", response_model=TransferDetail)
async def complete_transfer(
    transfer_id: str,
    house: House = Depends(get_house_member),
    db: AsyncSession = Depends(get_db),
):
    return await transfer_service.complete_transfer(transfer_id, house.id, db)


@router.post("/houses/{house_id}/transfers/{transfer_id}/containers", response_model=TransferDetail)
async def add_containers(
    transfer_id: str,
    body: AddContainersRequest,
    house: House = Depends(get_house_member),
    db: AsyncSession = Depends(get_db),
):
    return await transfer_service.add_containers(transfer_id, house.id, body.container_ids, db)


@router.delete(
    "/houses/{house_id}/transfers/{transfer_id}/containers/{container_id}",
    response_model=TransferDetail,
)
async def remove_container(
    transfer_id: str,
    container_id: str,
    house: House = Depends(get_house_member),
    db: AsyncSession = Depends(get_db),
):
    return await transfer_service.remove_container(transfer_id, container_id, house.id, db)


@router.get(
    "/houses/{house_id}/locations/{location_id}/summary",
    response_model=LocationSummaryOut,
)
async def location_summary(
    location_id: str,
    house: House = Depends(get_house_member),
    db: AsyncSession = Depends(get_db),
):
    return await transfer_service.get_location_summary(location_id, house.id, db)

File: frontend/src/stores/transfers.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright 2026 Lorenzo Benfenati
import { defineStore } from 'pinia'
import { ref } from 'vue'
import api from '@/services/api'
import type { ContainerSummary, LocationMini } from '@/stores/containers'

export interface TransferSummary {
  id: string
  name: string
  status: string
  destination_location: LocationMini | null
  scheduled_date: string | null
  vehicle_volume_liters: number | null
  container_count: number
  total_volume_liters: number
}

export interface TransferDetail extends TransferSummary {
  notes: string | null
  containers: ContainerSummary[]
  created_at: string
  updated_at: string
}

export interface TripPlan {
  trip_number: number
  containers: ContainerSummary[]
  total_volume_liters: number
  vehicle_fill_percent: number
  scheduled_date: string | null
}

export interface TransferPlanResponse {
  trips: TripPlan[]
  unassigned_containers: ContainerSummary[]
  algorithm: string
}

export interface LocationSummaryOut {
  location: LocationMini
  container_count: number
  total_volume_liters: number
  transferred_count: number
  planned_transfers: TransferSummary[]
  containers: ContainerSummary[]
}

export const useTransfersStore = defineStore('transfers', () => {
  const transfers = ref<TransferSummary[]>([])

  async function fetchTransfers(
    houseId: string,
    params?: { status?: string; destination_location_id?: string },
  ): Promise<TransferSummary[]> {
    const resp = await api.get<TransferSummary[]>(`/houses/${houseId}/transfers`, { params })
    transfers.value = resp.data
    return resp.data
  }

  async function fetchTransfer(houseId: string, transferId: string): Promise<TransferDetail> {
    const resp = await api.get<TransferDetail>(`/houses/${houseId}/transfers/${transferId}`)
    return resp.data
  }

  async function createTransfer(
    houseId: string,
    data: {
      name: string
      destination_location_id: string
      scheduled_date?: string | null
      vehicle_volume_liters?: number | null
      notes?: string | null
      container_ids?: string[]
    },
  ): Promise<TransferDetail> {
    const resp = await api.post<TransferDetail>(`/houses/${houseId}/transfers`, data)
    return resp.data
  }

  async function updateTransfer(
    houseId: string,
    transferId: string,
    data: Partial<{
      name: string
      destination_location_id: string
      scheduled_date: string | null
      vehicle_volume_liters: number | null
      notes: string | null
    }>,
  ): Promise<TransferDetail> {
    const resp = await api.put<TransferDetail>(`/houses/${houseId}/transfers/${transferId}`, data)
    return resp.data
  }

  async function startTransfer(houseId: string, transferId: string): Promise<TransferDetail> {
    const resp = await api.post<TransferDetail>(
      `/houses/${houseId}/transfers/${transferId}/start`,
    )
    return resp.data
  }

  async function completeTransfer(houseId: string, transferId: string): Promise<TransferDetail> {
    const resp = await api.post<TransferDetail>(
      `/houses/${houseId}/transfers/${transferId}/complete`,
    )
    return resp.data
  }

  async function addContainers(
    houseId: string,
    transferId: string,
    containerIds: string[],
  ): Promise<TransferDetail> {
    const resp = await api.post<TransferDetail>(
      `/houses/${houseId}/transfers/${transferId}/containers`,
      { container_ids: containerIds },
    )
    return resp.data
  }

  async function removeContainer(
    houseId: string,
    transferId: string,
    containerId: string,
  ): Promise<TransferDetail> {
    const resp = await api.delete<TransferDetail>(
      `/houses/${houseId}/transfers/${transferId}/containers/${containerId}`,
    )
    return resp.data
  }

  async function planTransfers(
    houseId: string,
    data: {
      destination_house_id: string
      vehicle_volume_liters: number
      scheduled_dates?: string[]
    },
  ): Promise<TransferPlanResponse> {
    const resp = await api.post<TransferPlanResponse>(
      `/houses/${houseId}/transfers/plan`,
      data,
    )
    return resp.data
  }

  async function fetchLocationSummary(
    houseId: string,
    locationId: string,
  ): Promise<LocationSummaryOut> {
    const resp = await api.get<LocationSummaryOut>(
      `/houses/${houseId}/locations/${locationId}/summary`,
    )
    return resp.data
  }

  return {
    transfers,
    fetchTransfers,
    fetchTransfer,
    createTransfer,
    updateTransfer,
    startTransfer,
    completeTransfer,
    addContainers,
    removeContainer,
    planTransfers,
    fetchLocationSummary,
  }
})
